File: python/144_binary_tree_preorder_traversal.py
# https://leetcode.com/problemset/algorithms/
# Given a binary tree, return the preorder traversal of its nodes' values.
# For example:
# Given binary tree {1,#,2,3},
#    1
#     \
#      2
#     /
#    3
# return [1,2,3].

# Definition for a binary tree node.
# class TreeNode(object):
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution(object):
    def preorderTraversal(self, root):
        """
        :type root: TreeNode
        :rtype: List[int]
        """
        rst = []
        self.preorder(root, rst)
        return rst
        
    def preorder(self, root, result):
        if root is None:
            return
        result.append(root.val)
        self.preorder(root.left, result)
        self.preorder(root.right, result)

File: java/144_binary_tree_preorder_traversal.java
// https://leetcode.com/problems/binary-tree-preorder-traversal/

// Given a binary tree, return the preorder traversal of its nodes' values.
// For example:
// Given binary tree {1,#,2,3},
//    1
//     \
//      2
//     /
//    3
// return [1,2,3].

/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode(int x) { val = x; }
 * }
 */
public class Solution {
    public List<Integer> preorderTraversal(TreeNode root) {
        List<Integer> rst = new ArrayList<Integer>();
        preorder(root,rst);
        return rst;
    }
    public void preorder(TreeNode root, List<Integer> rst){
        if (root == null){
            return;
        }
        rst.add(root.val);
        preorder(root.left, rst);
        preorder(root.right, rst);
    }
}